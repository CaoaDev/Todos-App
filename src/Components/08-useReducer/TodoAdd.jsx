import React from 'react';
import { useForm } from '../../Hooks/useForm';

export const TodoAdd = ( { handleAddTodo, dispatch } ) => {
    const [ { description }, handleInputChange, reset ] = useForm ( {
        description:''
    } );

    const handleSubmit = ( e ) => {
        e.preventDefault();

        if ( description.trim().length <= 1){
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };
        handleAddTodo( dispatch, newTodo );
        reset();
    }
    return (
    <div className='form-input'>
        <h4> Agregar Tarea... </h4>
        <hr />
        <form onSubmit={ handleSubmit }>
            <input
                type='text'
                name='description'
                className='form-control'
                placeholder='Anotar...'
                autoComplete='off'
                value={ description }
                onChange={ handleInputChange }
            />
            <button 
                type='submit'
                className='btn btn-outline-primary mt-1 btn-block'
            > 
                Agregar
            </button>
        </form>
    </div>
    )
}