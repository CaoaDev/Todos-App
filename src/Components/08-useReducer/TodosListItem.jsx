import React from 'react'

export const TodosListItem = ({
  todo,
  index,
  handleDelete,
  handleToggle,
  dispatch
}) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p
        className={`${todo.done && 'complete'}`}
        onClick={() => handleToggle(dispatch, todo.id)}
      >
        {index + 1}. {todo.desc}
      </p>
      <button
        className="btn btn-outline-danger"
        onClick={() => handleDelete(dispatch, todo.id)}
      >
        Borrar
      </button>
    </li>
  )
}
