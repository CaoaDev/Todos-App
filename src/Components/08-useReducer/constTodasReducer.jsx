export const handleAddTodo = ( dispatch, newTodo ) => {
    dispatch ( {
        type: 'add',
        payload: newTodo
    } );
};

export const handleDelete = ( dispatch, todoId ) => {
    dispatch ( {
        type: 'delete',
        payload: todoId
    } );
};
// se puede hacer de otra forma pero es mas larga para mi
     // const action = {
     //     type: 'delete',
     //     playload: todoId
     // } 
     // dispatch ( action );
    
export const handleToggle = ( dispatch, todoId ) => {
    dispatch ( {
        type: 'toggle',
        payload: todoId
    } );
};