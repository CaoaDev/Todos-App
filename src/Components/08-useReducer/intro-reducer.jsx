const intialState = [ {
    id: 1,
    todo: 'Despensa',
    done: false
}];

const newTodo = {
    id: 1,
    todo: 'Estudiar',
    done: false
};

// action no es nombre afuerzas puede ser cualquier cosa.
// ak se suman las dos const para poder usarlas en reducer
const todoReducer = ( state=intialState, action ) => {
    if ( action?.type === 'Agregar' ){
        return [ ...state, action.payload ]
    }
    return state;
}

let todos = todoReducer();

const action = {
    type: 'Agregar',
    payload: newTodo
};

todos = todoReducer ( todos, action );

// console.log( todos );