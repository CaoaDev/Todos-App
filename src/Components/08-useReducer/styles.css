input {
  align-items: center;
  color: green!important;
}

.complete {
  text-decoration: line-through;
}

.list-group-item{
  color: indigo;
  align-items: center;
  display: flex;
  justify-content: space-between;
}
