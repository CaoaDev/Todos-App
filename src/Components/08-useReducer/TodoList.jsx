import React from 'react';
import { TodosListItem } from '../../Components/08-useReducer/TodosListItem'

export const TodoList= ( { todos, handleDelete, handleToggle, dispatch } ) => {
  return (
    <ul className='list-group list-group-flush' >
      {
      todos.map ( ( todo, index ) => (
        <TodosListItem 
          key={ todo.id }
          todo={ todo }
          index={ index }
          handleDelete={ handleDelete }
          handleToggle={ handleToggle }
          dispatch={ dispatch }
        />
      ))
      }
    </ul>
  )
}