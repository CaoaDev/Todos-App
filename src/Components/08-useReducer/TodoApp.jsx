import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import { handleAddTodo, handleDelete, handleToggle } from './constTodasReducer';

import '../08-useReducer/styles.css';

const init = () => {
    return JSON.parse(localStorage.getItem( 'todos' ) ) || [];
}
//     return[{
//         id: new Date().getTime(),
//         desc: '',
//         done: false  
//     }];

export const TodoApp = () => {
    const [ todos, dispatch ] = useReducer( todoReducer, {} , init  );

    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify( todos ) )
}, [ todos ]);
    
    // console.log( description );
return (
<div>
    <h1><center> Agenda de Tareas. ( { todos.length } )</center></h1>
    <hr />

    <div className='row'>

        <div className='col-7'/>
        {
            todos.length > 0
            ? 
            < TodoList 
                todos= { todos }
                handleDelete={ handleDelete }
                handleToggle={ handleToggle }
                dispatch={ dispatch }
            />
            :
            <p>No se encontró Información...</p>
        }

        <div className='col-5' />

        < TodoAdd
            handleAddTodo={ handleAddTodo }
            dispatch={ dispatch }
        />

        </div>
    </div>
    )
}