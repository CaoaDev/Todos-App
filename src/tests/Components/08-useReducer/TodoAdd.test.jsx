import React from 'react';
import { shallow } from 'enzyme';
import { TodoAdd } from '../../../Components/08-useReducer/TodoAdd';

describe('Prueba en <TodoAdd />', () => { 
    const mockEvent = {
        preventDefault: jest.fn()
    };
    const handleAddTodo=jest.fn();
    const dispatch=jest.fn();
    
    const wrapper=shallow(
    <TodoAdd
    handleAddTodo={ handleAddTodo } 
    dispatch={ dispatch }
    />
    );
    
    test('Debe mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('No debe mandar el handleAddTodo', () => {
        const formSubmit= wrapper.find( 'form' ).prop( 'onSubmit' );
        formSubmit( { preventDefault () {} } );
        expect( handleAddTodo ).toHaveBeenCalledTimes( 0 );
    });

    test('Debe llamar la funcion handleAddTodo', () => {
        const value= 'Aprender React';
        wrapper.find( 'input' ).simulate( 'change', {
            target:{
                value,
                name: 'description'
            }
        } );
        const formSubmit= wrapper.find( 'form' ).prop( 'onSubmit' );
        formSubmit( { preventDefault() {} } );
        expect( handleAddTodo ).toHaveBeenCalledTimes( 1 );
        expect( handleAddTodo ).toHaveBeenCalledWith( expect.any(Object) );
        expect( ( handleAddTodo ) ).toHaveBeenCalledWith({
            id: expect.any( Number ),
            des: value,
            done: false
        });
        expect( wrapper.find( 'input' ).prop( 'vale' ) ).toBe( '' );
    });
})