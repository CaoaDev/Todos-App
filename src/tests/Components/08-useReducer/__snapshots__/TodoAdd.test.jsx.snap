// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba en <TodoAdd /> Debe mostrarse correctamente 1`] = `
<div
  className="form-input"
>
  <h4>
     Agregar Tarea... 
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Anotar..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</div>
`;
